### PR TITLE
[WIP] Create delete basic integration test

### PR DESCRIPTION
**Pull request by spew**
**What this PR does / why we need it**:
This PR adds a basic integration test which creates and deletes a cluster. Note to interested parties, please ignore the first commit in this change "Implement clusterctl delete cluster" as that is a copy of another, dependent PR that has not yet merged.

**Release note**:
```release-note
NONE
```

@kubernetes/kube-deploy-reviewers

### CODE REVIEW

**Comment by ingvagabund** [comment]
@spew https://github.com/kubernetes-sigs/cluster-api/pull/406 is merged. Does it still make sense to rebase the PR and introduce the integration testing for the create->delete scenario?

File: clusterctl/clusterdeployer/clusterdeployer.go [hunk 1]
```diff
@@ -388,13 +451,45 @@ func pivot(from, to ClusterClient) error {
...
 		if err != nil {
-			return err
+			return fmt.Errorf("error moving Machine '%v': %v", machine.GetName(), err)
 		}
 		glog.Infof("Moved Machine '%s'", machine.GetName())
 	}
 	return nil
 }
 
+func deleteObjects(client ClusterClient) error {
```
Comment (frobware) [inline clusterctl/clusterdeployer/clusterdeployer.go]: Wondering two things 1) if I get an `error` can I call it again and 2) from a readability perspective would it be better to split it up into 3 helper functions: (`DeleteMachineDeploymentObjects`, `DeleteMachineSetObjects`, `DeleteMachineObjects`).

File: clusterctl/clusterdeployer/clusterdeployer.go [hunk 2]
```diff
@@ -440,3 +535,10 @@ func containsMasterRole(roles []clustercommon.MachineRole) bool {
 	}
 	return false
 }
+
+func closeClient(client ClusterClient, name string) {
```
Comment (frobware) [inline clusterctl/clusterdeployer/clusterdeployer.go]: Let this return `error` and let the caller decide what to do with it, if anything.

File: clusterctl/clusterdeployer/clusterdeployer_test.go [hunk 3]
```diff
@@ -109,24 +125,66 @@ func (c *testClusterClient) GetClusterObjects() ([]*clusterv1.Cluster, error) {
 	return c.clusters, c.GetClusterObjectsErr
 }
 
+func (c *testClusterClient) GetMachineDeploymentObjects() ([]*clusterv1.MachineDeployment, error) {
```
Comment (frobware) [inline clusterctl/clusterdeployer/clusterdeployer_test.go]: Should this return a copy? What happens if, after the callers holds on to it for a while, it is mutated? Ditto for `GetMachineSetObjects()`.

File: clusterctl/clusterdeployer/clusterdeployer_test.go [hunk 4]
```diff
@@ -109,24 +125,66 @@ func (c *testClusterClient) GetClusterObjects() ([]*clusterv1.Cluster, error) {
...
+}
+
 func (c *testClusterClient) GetMachineObjects() ([]*clusterv1.Machine, error) {
 	return c.machines, c.GetMachineObjectsErr
 }
 
 func (c *testClusterClient) CreateClusterObject(cluster *clusterv1.Cluster) error {
-	if c.CreateClusterObjectErr != nil {
-		return c.CreateClusterObjectErr
+	if c.CreateClusterObjectErr == nil {
```
Comment (frobware) [inline clusterctl/clusterdeployer/clusterdeployer_test.go]: Why not handle the error case first (though admittedly I could be misreading the github diff):

    if c.CreateClusterObjectErr != nil {
        return c.CreateClusterObjectErr
    }
    c.clusters = append(c.clusters, cluster)
    return nil

**Comment by spew** [comment]
Hi @frobware thanks for the review. Most of those changes have already been merged (this is a two commit PR and the initial commit was reviewed seperately).

**Comment by frobware** [comment]
> thanks for the review. Most of those changes have already been merged (this is a two commit PR and the initial commit was reviewed seperately).

:)

OK, mention me if you want a review on the remaining part(s)...

**Comment by spew** [comment]
Just rebased through github -- I think the tests will not pass yet due to how the prow testing infra works.

**Comment by frobware** [comment]
> Just rebased through github -- I think the tests will not pass yet due to how the prow testing infra works.

Not sure if you actually meant to `rebase` (and whether you care) but it reports as a giant `merge` for me.

**Comment by roberthbailey** [comment]
This PR has been open for 8 months and hasn't been updated for the last 7 months. I've been leaving it open with the hope that we'd still add an integration test into the repo but when we do I no longer think that this PR will be the starting point for that work. 

Thanks @spew for pulling this together and I'm sorry that we weren't able to get it merged. 

/close

**Pull request closed without merging**